src/libexpr/primops/fetchTree.cc

```diff
@@ -48,10 +51,40 @@ void emitTreeAttrs(
     v.attrs->sort();
 }
 
-static void prim_fetchTree(EvalState & state, const Pos & pos, Value * * args, Value & v)
+std::string fixURI(std::string uri, EvalState &state)
+{
+    state.checkURI(uri);
+    return uri.find("://") != std::string::npos ? uri : "file://" + uri;
+}
+
+void addURI(EvalState &state, fetchers::Attrs &attrs, Symbol name, std::string v)
+{
+    string n(name);
+    attrs.emplace(name, n == "url" ? fixURI(v, state) : v);
+}
+
+void validateFetcherAttrs(fetchers::Attrs attrs, const Pos &pos)
 {
-    settings.requireExperimentalFeature("flakes");
+    if (!attrs.count("type"))
+        throw Error("attribute 'type' is missing in call to 'fetchTree', at %s", pos);
+
+    if (getStrAttr(attrs, "type") == "git") {
+        auto narHash = maybeGetStrAttr(attrs, "narHash");
+        auto ref = maybeGetStrAttr(attrs, "ref");
+        auto rev = maybeGetStrAttr(attrs, "rev");
+
+        if (narHash && !ref && !rev)
```

Comment:
This validation should be done in `git.cc`, not `fetchTree.c`. Also, I think this test isn't really necessary, since people should be free to specify a narHash without a rev/ref. It could be fine in flake.nix, since the lock file will contain the rev...